starfyre/parser.py

```diff
@@ -107,35 +108,30 @@ def handle_starttag(self, tag, attrs):
             uuid=uuid4(),
         )
 
-        # instead of assiging tags we assign uuids
-        self.stack.append((component, self.current_depth))
-        [(element[0].tag, element[1]) for element in self.stack]
+        # instead of assiging tags we assign uuids        
+        
+        self.stack.append(component)
+        
 
     def handle_endtag(self, tag):
         # we need to check if the tag is a default component or a custom component
-        # if it is a custom component, we get the element from the custom components dict
+        # if it is a custom component, we get the element from the custom components dict          
+
         if tag not in self.generic_tags and tag in self.components:
             component = self.components[tag]
             tag = component.tag
-
-        # need to check the if this is always true
-        parent_node, parent_depth = self.stack[
-            -1
-        ]  # based on the assumption that the stack is not empty
-
-        while len(self.children) > 0:
-            child, child_depth = self.children[0]
-            if child_depth == parent_depth + 1:
-                self.children.pop(0)
-                self.stack[-1][0].children.insert(0, child)
-            else:
-                break  # we have reached the end of the children
-
-        self.stack.pop()
+        
+        endtag_node = self.stack.pop()  
         self.current_depth -= 1
+        if endtag_node.tag != "style" and endtag_node.tag != "script":
+            if len(self.stack) > 0:
+                parent_node = self.stack[-1]      #this is last item/"top element" of stack
+                parent_node.children.append(endtag_node)
+            else:
+                self.root_node = endtag_node
+      
```

Comment:
<Can you please explain what was the purpose of removing self.children and depth. i.e. can you please provide an overview of the new algorithm?>
@sansyrox 
I used stack structure to track hierarchy between HTML tags - when tag is opened, it is added to our stack, (reference line 95 of parse.py).
When tag is closed, the component for this tag is removed from our stack, added to children list of its parent node, and the next element in our stack will be our closing node parent “top element on stack”.
So we don't need to have separate RootParser.children list, and we don't need to track depth of HTML tags anymore - we got this by using RootParser.stack. When element is removed from stack - it added to children list of it's parent node, and if it is root element meaning last element remaining in RootParser.stack , we just remember it in root_node attribute.